source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,305 @@
+    if (*iter == SEPARATOR) {
+      ++iter;
+    }
+
+    // Move our last marker to the first position after where we stopped. Search for the next name
+    // separator
+    last += count;
+    ++last;
+    count = name_.find_first_of(SEPARATOR, last);
+  }
+
+  // Write the remaining segment prepended by its length
+  count = name_.size() - last;
+  buffer_.writeBEInt<uint8_t>(count);
+  for (size_t i = 0; i < count; i++) {
+    buffer_.writeByte(*iter);
+    ++iter;
+  }
+
+  // Terminate the name record with a null byte
+  buffer_.writeByte(0x00);
+}
+
+// Serialize a DNS Query Record
+void DnsQueryRecord::serialize(Buffer::OwnedImpl& output) {
+  buffer_.drain(buffer_.length());
+
+  serializeName();
+  buffer_.writeBEInt<uint16_t>(type_);
+  buffer_.writeBEInt<uint16_t>(class_);
+
+  output.add(buffer_);
+}
+
+bool DnsMessageParser::parseDnsObject(const Buffer::InstancePtr& buffer) {
+
+  auto available_bytes = buffer->length();
+
+  memset(&incoming_, 0x00, sizeof(incoming_));
+
+  static constexpr uint64_t field_size = sizeof(uint16_t);
+  uint64_t offset = 0;
+  uint16_t data;
+
+  DnsQueryParseState state_{DnsQueryParseState::Init};
+
+  while (state_ != DnsQueryParseState::Finish) {
+
+    // Ensure that we have enough data remaining in the buffer to parse the query
+    if (available_bytes < field_size) {
+      ENVOY_LOG(error,
+                "Exhausted available bytes in the buffer. Insufficient data to parse query field.");
+      return false;
+    }
+
+    // Each aggregate DNS header field is 2 bytes wide.
+    data = buffer->peekBEInt<uint16_t>(offset);
+    offset += field_size;
+    available_bytes -= field_size;
+
+    if (offset > buffer->length()) {
+      ENVOY_LOG(error, "Buffer read offset [{}] is beyond buffer length [{}].", offset,
+                buffer->length());
+      return false;
+    }
+
+    switch (state_) {
+    case DnsQueryParseState::Init:
+      incoming_.id = data;
+      state_ = DnsQueryParseState::Flags;
+      break;
+
+    case DnsQueryParseState::Flags:
+      ::memcpy(static_cast<void*>(&incoming_.flags), &data, sizeof(uint16_t));
+      state_ = DnsQueryParseState::Questions;
+      break;
+
+    case DnsQueryParseState::Questions:
+      incoming_.questions = data;
+      state_ = DnsQueryParseState::Answers;
+      break;
+
+    case DnsQueryParseState::Answers:
+      incoming_.answers = data;
+      state_ = DnsQueryParseState::Authority;
+      break;
+
+    case DnsQueryParseState::Authority:
+      incoming_.authority_rrs = data;
+      state_ = DnsQueryParseState::Authority2;
+      break;
+
+    case DnsQueryParseState::Authority2:
+      incoming_.additional_rrs = data;
+      state_ = DnsQueryParseState::Finish;
+      break;
+
+    case DnsQueryParseState::Finish:
+      break;
+
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+  // Verify that we still have available data in the buffer to read answer and query records
```

Comment:
you mean read only query records right? Not answer

Reply:
I'm actually using this function to parse the response to verify data in tests.   Ideally yes, the answers record count should be zero.  We can enforce this with the flags to verify that queries don't have answers, but I think this is rare.  I will add a TODO.